### PR TITLE
[JENKINS-63584] Add an option to use Jenkins proxy to access Kubernetes

### PR DESCRIPTION
**Pull request by inidona**

### CODE REVIEW

**Review by Vlatombe — CHANGES_REQUESTED** [review]
This should be an option at the Kubernetes Cloud level. If the master is deployed in the same Kubernetes cluster as the agents it will provision, this would cause it to use the configured proxy instead of targetting directly the Kubernetes API.

**Comment by inidona** [comment]
> This should be an option at the Kubernetes Cloud level. If the master is deployed in the same Kubernetes cluster as the agents it will provision, this would cause it to use the configured proxy instead of targetting directly the Kubernetes API.

Do you mean a checkbox at cloudconfig that jenkins proxy should be used ?

**Comment by Vlatombe** [comment]
Yes.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java [hunk 1]
```diff
@@ -29,16 +29,19 @@
 import java.util.HashMap;
 import java.util.Map;
 
+import jenkins.model.Jenkins;
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java:32]: Remove unused imports

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java [hunk 2]
```diff
@@ -29,16 +29,19 @@
...
 import org.junit.After;
 import org.junit.Before;
 import org.junit.Test;
 
 import static io.fabric8.kubernetes.client.Config.*;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import org.junit.runner.RunWith;
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java:39]: Ditto

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java [hunk 3]
```diff
@@ -29,16 +29,19 @@
...
 
 /**
  * Test the creation of clients
  */
+
```
Comment (Vlatombe) [inline src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java:44]: Revert

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapter.java [hunk 4]
```diff
@@ -120,9 +124,38 @@ public KubernetesClient createClient() throws KubernetesAuthException {
         }
 
         LOGGER.log(FINE, "Creating Kubernetes client: {0}", this.toString());
+        // JENKINS-63584 If Jenkins has an configured Proxy and cloud has enabled proxy usage pass the arguments to K8S
+        LOGGER.log(FINE, "Proxy Settings for Cloud: " + useJenkinsProxy);
+        if(useJenkinsProxy) {
+            Jenkins jenkins = Jenkins.getInstanceOrNull(); // this code might run on slaves
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapter.java:130]: Not really.

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapter.java [hunk 5]
```diff
@@ -120,9 +124,38 @@ public KubernetesClient createClient() throws KubernetesAuthException {
...
+            LOGGER.log(FINE, "Jenkins Instance: " + jenkins);
+            if (jenkins != null) {
+                ProxyConfiguration p = jenkins.proxy;
+                LOGGER.log(FINE,"Proxy Instance: " + p);
+                if (p != null) {
+                    builder.withWebsocketTimeout(10000);
```
Comment (Vlatombe) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapter.java:136]: Why this?

**Review by Vlatombe — COMMENTED** [review]
Some minor reviews and a question. Otherwise, this lgtm.

File: src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java (hunk 3 above)
Comment (inidona) [reply src/test/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapterTest.java]: class reverted

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapter.java (hunk 4 above)
Comment (inidona) [reply src/main/java/org/csanchez/jenkins/plugins/kubernetes/KubernetesFactoryAdapter.java]: comment removed

**Review by Vlatombe — APPROVED** [review]

**Comment by pashcan** [comment]
Greetings,
Why does [JENKINS-63584](https://issues.jenkins.io/browse/JENKINS-63584) show **Fixed but Unreleased** if this was merged many moons ago?
I am actually seeing configured proxy being ignored when trying to reach remote cluster, so I'm wondering if this in fact made it into plugin. Despite having toggled Use Jenkins Proxy checkbox, I am seeing in the logs ```org.csanchez.jenkins.plugins.kubernetes.KubernetesFactoryAdapter
Proxy Settings for Cloud: false```

**Comment by inidona** [comment]
@pashcan i don´t know why the issue is not resolved, but the release works for me as expected. 
I found out for  communication we had to use websockets and the proxy and maybe the reverse proxy before your k8s cluster must support websockets too.

**Comment by duemir** [comment]
@pashcan it looks like "Test Connection" was not updated so it doesn't take into account the "Jenkins Proxy" flag defaulting to `false`, i.e. don't use.

**Comment by duemir** [comment]
I tried to fix "Test Connection" in #1064

**Pull request merged**